Prototype topic generation using gpt-3.5-turbo-16k

kathy-t opened the pull request:
**Description**
This PR adds a topicgenerator program that generates topics using OpenAI's GPT-3.5-turbo-16k AI model. The program uses the [Chat Completions API](https://platform.openai.com/docs/guides/gpt/chat-completions-api) to ask the AI model to summarize the contents of the main descriptor file. Unfortunately, we can't send everything because the AI model has a [token](https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them#h_4ae02e0c88) limit. The `gpt-3.5-turbo-16k` model has a limit of 16,385 tokens. You can see the limits of the various models [here](https://platform.openai.com/docs/models/overview).

I chose to use `gpt-3.5-turbo-16k` over `gpt-3.5-turbo` and `gpt-4` because it has a greater token limit. `gpt-3.5-turbo` has a limit of 4,097 tokens and `gpt-4` has a limit of 8,192 tokens. I also chose `gpt-3.5-turbo-16k` over the more powerful `gpt-4` because it's faster, has a higher rate limit, and cheaper (can see pricing per token [here](https://openai.com/pricing)). In my experimentation, `gpt-4` was much slower, and it kept running into time-outs.

For the purposes of the prototype, this program takes in a CSV file that contains the TRS ID and version name of the entries to generate topics for. When we want to run through all the entries, we'll have to modify this program, but for now, I think this is okay while we evaluate the topics that it generates.

The program outputs a CSV file containing the AI topics, as well as other information like the number of tokens used, whether or not the descriptor file was truncated, etc.

**Results**
I ran this program on 23 entries, specified in the entries.csv file. I selected the most starred workflows, some of Ash's workflows, a notebook, and a tool. Feel free to suggest more workflows to add! The results are in the `results` directory. 

I did add a CSV file of topics generated using GPT4 so people can compare the topics from GPT 3.5-turbo and GPT 4. Note how not every entry in entries.csv has a corresponding row in the GPT4 CSV file. This is because the request timed out. It also took much longer to run.

The program as is in the PR does not use GPT4, but the PR can easily be modified to do so. What do others think about GPT 3.5 vs GPT 4? IMO, I think GPT 3.5 does just fine for our purpose, but I'm not a workflow developer :slightly_smiling_face: 

**Review Instructions**
Not sure if reviewable by others unless they have an OpenAI API key.

**Issue**
[SEAB-5907](https://ucsc-cgl.atlassian.net/browse/SEAB-5907)

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install` in the project that you have modified (until https://ucsc-cgl.atlassian.net/browse/SEAB-5300 adds multi-module support properly)
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If you are changing dependencies, check with dependabot to ensure you are not introducing new high/critical vulnerabilities
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead. 


[SEAB-5907]: https://ucsc-cgl.atlassian.net/browse/SEAB-5907?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kathy-t commented on topicgenerator/results/generated-topics_GPT_3_5_TURBO_16K_20231102T195407Z.csv:
This is the generated topics using gpt-3.5-turbo-16k - this PR uses this AI model

kathy-t commented on topicgenerator/results/generated-topics_GPT_4_20231102T210454Z.csv:
This is the generated topics using gpt-4. This PR currently does not use this model, but could be modified to do so. This file is here for reviewers to compare the quality of the topics to the gpt-3.5 versions.

svonworl commented:
This is cool!  Some random thoughts:

1. The cost is linear with the number of tokens, and, in this case, we submit a relatively large query Q, composed of our ask A and the primary descriptor source S, and the ai responds with a short response R.  If that's true, and |Q| >> |A| + |R|, we could make additional asks with short responses for very small incremental cost.  Some ideas for additional asks:
* "Make that shorter." (to get below 140 chars)
* "Make that more technical." / "Rephrase that for the layperson."
* "Convert that to a verb phrase." (to eliminate the "This workflow does" prefix)
* "Generate a list of search keywords."
* "What data formats does it use?"
* "Who are the authors?"
* "What institution(s) are they with?"
Even if we don't anticipate immediately using the results, the extra questions are almost free.  We could warehouse them for possible later use...
2. Would be interesting to see how well it condenses a description into a topic sentence, both with and without the accompanying primary descriptor source.
3. Potentially, could condense the input to save cost.  For example, we could filter the images from the notebook json.  Might even be able to use a much faster, cheaper "assistant" model to preprocess the input...

svonworl commented:
I like the GPT-3.5 topic sentences a little better but I'm not sure why.

svonworl commented:
> This is cool! Some random thoughts:

(None of this needs to happen in this PR).

coverbeck commented on topicgenerator/README.md:
```diff
@@ -0,0 +1,62 @@
+# Topic Generator
+
+This is a Java program that generates topics for public Dockstore entries and puts the results into the webservice using OpenAI's gpt-3.5-turbo-16k AI model.
```
Nit: it doesn't put them into the web service (yet?), it generates a file.

coverbeck reviewed (approved):
Nice work! I'm very impressed with the generated topic sentences, albeit coming from the perspective of somebody who barely knows any bioinformatics -- but they sound good to me. :)

> "Convert that to a verb phrase." (to eliminate the "This workflow does" prefix)

I strongly second that suggestion by Steve -- if we end up displaying these topic sentences, e.g., in search, the "This workfow..." would get very repetitive.

svonworl commented:
Another idea; we'll probably regenerate/refresh this information someday, so it might make sense to have a mechanism to tell if the input has changed, so we can skip the unchanged entries (if it makes sense).  Maybe store a hash of the input (in this case, the primary descriptor) in the output?

kathy-t commented:
>     1. Per my [limited] understanding, the cost is linear with the number of tokens, and, in this case, we submit a relatively large query, composed of our ask A and the primary descriptor source S, and the ai responds with a response R.  If that's true, and |S| >> |A| + |R|, we could make additional short asks with short responses for very small incremental cost.  Some ideas for additional asks:
> 
> 
>     * "Make that shorter." (to get below 140 chars)
> 
>     * "Make that more technical." / "Rephrase that for the layperson."
> 
>     * "Convert that to a verb phrase." (to eliminate the "This workflow does" prefix)
> 
>     * "Generate a list of search keywords."
> 
>     * "What data formats does it use?"
> 
>     * "Who are the authors?"
> 
>     * "What institution(s) are they with?"
>       Even if we don't anticipate immediately using the results, the extra asks are almost free.  We could warehouse them for possible later use...

The extra asks are not quite free. In order to do extra asks, we have to send the entirety of the previous conversation, including the initial "Summarize this workflow" ask and its response, which chews into the context limit for the extra asks. One way to get around that is to summarize the previous conversation then submit the extra ask, but I think this is something we can do in a follow-up ticket if we want to go down this route.

kathy-t commented:
Updated the request so that the sentence starts with a verb and re-generated the GPT-3.5-turbo topics. Removed the GPT-4 topic results because it seems like we're okay with GPT-3.5-turbo

svonworl commented:
> > ```
> > 1. Per my [limited] understanding, the cost is linear with the number of tokens, and, in this case, we submit a relatively large query, composed of our ask A and the primary descriptor source S, and the ai responds with a response R.  If that's true, and |S| >> |A| + |R|, we could make additional short asks with short responses for very small incremental cost.  Some ideas for additional asks:
> > 
> > 
> > * "Make that shorter." (to get below 140 chars)
> > 
> > * "Make that more technical." / "Rephrase that for the layperson."
> > 
> > * "Convert that to a verb phrase." (to eliminate the "This workflow does" prefix)
> > 
> > * "Generate a list of search keywords."
> > 
> > * "What data formats does it use?"
> > 
> > * "Who are the authors?"
> > 
> > * "What institution(s) are they with?"
> >   Even if we don't anticipate immediately using the results, the extra asks are almost free.  We could warehouse them for possible later use...
> > ```
> 
> The extra asks are not quite free. In order to do extra asks, we have to send the entirety of the previous conversation, including the initial "Summarize this workflow" ask and its response, which chews into the context limit for the extra asks. One way to get around that is to summarize the previous conversation then submit the extra ask, but I think this is something we can do in a follow-up ticket if we want to go down this route.

Bummer, this news makes me frowny (not at you, Kathy, at the API).  Seems unfair to get billed over and over again for repeated input tokens that represent the "current session" state, we already paid for them.  Oh well.

If we did ever want to submit multiple asks, another approach might be to submit all the asks in the same prompt, something like:
Given the following information, do A, B, C, and D, and separate each result by X. <information>
